README.md
# angular-momentum-scroll

AngularJS directive that adds momentum scroll via iScroll (http://cubiq.org).

## Demo
Find a demonstration on http://wzr1337.github.com

## Usage

* Include `scrollable.js` or `scrollable.min.js` into your page
* Declare `'angular-momentum-scroll'` as a dependency for your angular app: `angular.module('myApp', ['angular-momentum-scroll']);`
* Add the `scrollable` attribute to a container of your choice:
    * the container MUST have set width and height !!
    * in order to scroll horizontally the .scroller class inside of your container MUST have set width and height !!
* The container takes an additional attribute `parameters` according to iScroll docs. e.g. 
   
    ```javascript
    <div id="my-cont" style="height: 400px; width: 100%;" scrollable 
    parameters="{{ {vScroll : true, snap: '.row'} }}">

* You can register any of the following handlers by specifying a callback function via attribute:
    * onRefresh
    * onBeforeScrollStart
    * onScrollStart
    * onBeforeScrollMove
    * onScrollMove
    * onBeforeScrollEnd
    * onScrollEnd
    * onTouchEnd
    * onDestroy
    * onZoomStart
    * onZoom
    * onZoomEnd

    ```javascript
    <div id="my-cont" style="height: 400px; width: 100%;" scrollable
    parameters="{{ {scrollbars : true, snap: '.row'} }}"
    on-scroll-move="doSometginhg()">...</div>

* You can bind a variable to the curr-page-x and curr-page-y attribute. The data binding is bidirectional, so that you can scroll programatically as well as being notified on page change.

  ```javascript
    <div id="my-cont" style="height: 400px; width: 100%;" scrollable
    parameters="{{ {vScroll : true, snap: '.row'} }}"
    curr-page-y="myPageY">...</div>

* You can watch isMinY, isMinX and isMaxX, isMAxY to find out, if the scrlller reached its final top or bottom, left or right position. These help you to build an infinitescroll or similar. use them as follows:

  ```javascript
  <ol id='mycontainer' scrollable
    parameters="{{ {vScroll:true, scrollbars: true, snap: 'li'} }}"
    curr-page-x="currPageX" curr-page-y="currPageY" curr-x="currX"
    curr-y="currY" is-min-y="isMinY" is-max-y="isMaxY">
    <li class="listitem" ng-repeat="thing in awesomeThings">{{thing}}</li>
  </ol>


changelog.md
# Change-Log

## 2016-03-20 v0.5.11
properly cleaning up event handler and destroy potential outdated instance before init

## 2014-09-25
fixed click behavior for older webkit browsers

## 2014-9-24
* Listening to resize of scroller instead of number of elements
* Default option preventDefault set to false (allow clickthrough)

## 2014-07-14
* Moved to version number 0.5.8
* Fixed bower.json of distribution package so that the ```main``` path is now a valid path
    * Fixes issue that grunt-wiredep couldn't inject angular-momentum-scroll
    * Bower package version contains now build timestamp
* Modified build process so that the minified file holds now proper license informations of angular-momentum-scroll and the used iScroll library
* Wrapped scrollable.js into an IIFE to avoid code leaking
* Source map gets now included into the distribution package
* Updated NPM and Bower dependencies
* Updated Karma test suite

## 2014-06-15
fixed check for scrollX

## 2014-01-17
fixed horizontal scroll css

## 2013-12-18
fixed $destroy

## 2013-11-30
fixed reinitialization on DOM extension
fixed accessing undefined scope vars

## 2013-10-31
fixed horizontal scroll for new iScroll parameter scrollX

## 2013-10-30
Fixed regular scroll

## 2013-10-18
migrated to iscroll 5 for better performance on Android

## 2013-09-19 v0.4.9
initialization of snappping scroll fixed

## 2013-09-16 v0.4.8
smaller fix for orientation change

## 2013-09-16 v0.4.7


## 2013-08-27 v0.4.6
Fixed horizontal scroll (css)

## 2013-06-09v0.4.4
Add isMaxX/isMinX/isMaxY/isMinY as an interface to the final TOP and bottoms positions.

## 2013-06-08 v0.4.3
Now supports X and Y coordinates of scroller

## 2013-05-17 v0.4.1
Fixed a bug with emtpy scrollers

## 2013-05-16 v0.4.0
Added two way data bound pages

## 2013-05-09 v0.3.0
Added callbacks for 'on' handles of iScroll

## 2013-05-08 v0.2.0
now supporting scrollToPage method of iScroll

## 2013-05-06 v0.1.1
Added some css hack to allow horizontal scrolling with css only. No hacky width calculations necessary


test/spec/scrollable.js
'use strict';

describe('Directive: scrollable', function () {
  beforeEach(module('angular-momentum-scroll'));

  var element, scope;

  it('should make hidden element visible', inject(function ($rootScope, $compile) {
    element = angular.element('<div scrollable parameters={{{foo:\'bar\' }}}><ul>' +
            '<li>Item1</li><li>Item2</li>' +
            '<li>Item3</li></ul></div>');
    scope = $rootScope.$new();
    element = $compile(element)($rootScope);
    expect(scope.parameters).toBe(undefined);
  }));
});


demo/test/spec/controllers/main.js
'use strict';

describe('Controller: MainCtrl', function () {

  // load the controller's module
  beforeEach(module('demoApp'));

  var MainCtrl,
    scope;

  // Initialize the controller and a mock scope
  beforeEach(inject(function ($controller, $rootScope) {
    scope = $rootScope.$new();
    MainCtrl = $controller('MainCtrl', {
      $scope: scope
    });
  }));

  it('should attach a list of awesomeThings to the scope', function () {
    expect(scope.awesomeThings.length).toBe(40);
  });

  it('should attach a current values equal to 0 to the scope', function () {
    expect(scope.currX).toBe(0);
    expect(scope.currY).toBe(0);
    expect(scope.currPageX).toBe(0);
    expect(scope.currPageY).toBe(0);
    expect(scope.isMaxY).toBe(false);
    expect(scope.isMinY).toBe(false);
    expect(scope.iScrollParameters).toEqual(jasmine.any(Object));
  });

  it('should append an item to the of awesomeThings on the scope', function () {
    var _len = scope.awesomeThings.length;
    scope.append('appended item');
    expect(scope.awesomeThings.length).toBe(_len + 1);
    expect(scope.awesomeThings[_len]).toEqual('appended item');
  });

  it('should prepend an item to the of awesomeThings on the scope', function () {
    var _len = scope.awesomeThings.length;
    scope.prepend('prepended item');
    expect(scope.awesomeThings.length).toBe(_len + 1);
    expect(scope.awesomeThings[0]).toEqual('prepended item');
  });
});


Gruntfile.js
'use strict';

module.exports = function(grunt) {

  var isPreReleaseBuild = grunt.option('pre-release') !== undefined; 

  // load all grunt tasks
  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    buildversion: (isPreReleaseBuild) ? '<%= pkg.version %>-' + grunt.template.today('yymmddHHMM') : '<%= pkg.version %>',
    clean : {
      dist : {
        files : [ {
          dot : true,
          src : [ 'dist/**' ]
        } ]
      },
      server : '.tmp'
    },
    ngAnnotate: {
      options: {
        singleQuotes: true,
      },
      dist: {
        files : {
          'dist/scrollable.js' : [
            'bower_components/iscroll/build/iscroll.js',
            'src/scrollable.js'
          ]
        }
      },
    },
    uglify: {
      options: {
        // Keep license informations
        preserveComments: 'some',
        sourceMap: true
      },
      dist: {
        files: { 'dist/scrollable.min.js': [ 'dist/scrollable.js' ] }
      }
    },
    jshint : {
      options : {
        jshintrc : '.jshintrc'
      },
      all : [ 'Gruntfile.js', 'src/*.js' ]
    },
    karma : {
      unit : {
        configFile : 'karma.conf.js',
        singleRun : true
      }
    },
    copy: {
      main: {
        files: [
          {expand: true, src: ['./*.md'], dest: 'dist/' , filter: 'isFile'}, // copy *.md
        ]
      }
    },
    'json-replace': {
      options: {
        space: '  ',
      },
      bowerdist: {
        options: {
          replace: {
            version: '<%= buildversion %>',
            devDependencies: '',
            scripts : '',
            homepage: 'https://github.com/wzr1337/bower-angular-momentum-scroll',
            repository: {
              'type': 'git',
              'url': 'git://github.com/wzr1337/bower-angular-momentum-scroll'
            },
            main: './scrollable.min.js'
          }
        },
        files: {'dist/bower.json': [ 'bower.json' ]}
      },
      bower: {
        options: {
          replace: {
            version: '<%= pkg.version %>'
          }
        },
        files: {'bower.json': [ 'bower.json' ]}
      }
    }
  });

  grunt.registerTask('build', [ 'clean:dist',
                                'jshint',
                                'ngAnnotate',
                                'uglify',
                                'copy',
                                'json-replace']);

  grunt.registerTask('test', ['karma']);

  grunt.registerTask('default', [ 'build' ]);
};


karma.conf.js
// Karma configuration

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jasmine'],


    // list of files / patterns to load in the browser
    files: [
        'bower_components/angular/angular.js',
        'bower_components/angular-mocks/angular-mocks.js',
        'bower_components/iscroll/build/iscroll.js',
        'src/scrollable.js',
        'test/spec/**/*.js'
    ],


    // list of files to exclude
    exclude: [
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
    },


    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],


    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};


src/scrollable.js
/**
 * @license angular-momentum-scroll
 * (c) 2012-2014 Patrick Bartsch
 * Momentum scroll for AngularJS based on iScroll
 * License: MIT
 */
(function(angular) {

'use strict';

angular.module('angular-momentum-scroll', []);

/**
 * scrollable directive utilizing iScroll lib. Attach to wrapping container.
 * First child will be scrolled only, SEE: http://cubiq.org/iscroll-4
 *
 * automatically sets following CSS style on the container: overflow: auto;
 * position: relative;
 *  !! the container MUST have set width and height !!
 *  !! in order to scroll horizontally the .scroller class inside of your
 * container MUST have set width and height !!
 *
 * the container takes an additional attribute 'iscroll-parameters' according to
 * iScroll docs.
 *
 * e.g. <div id="my-cont" style="height: 400px; width: 100%;" scrollable
 * parameters="{{ {hScrollbar : true, snap: '.row'} }}">...</div>
 *
 * You can bind a page number from the parents container to the scrollers
 * current page by using curr-page-x and curr-page-y attributes. The binding is
 * two way, so your parent scope will be updated also on scrolling.
 * This means, you can use curr-page-x and curr-page-y for scrolling to a page 
 * programatically as well as for listening to page changes. 
 *
 * <div scrollable style="height: 80%; width: 100%;" 
 *      curr-page-y="{{ mypage }}" scroll-to-page-time = "5000"
 *      parameters="{{ { hScroll: true, hScrollbar: false, snap: true,
 *      momentum: false} }}"">...</div>
 */
angular.module('angular-momentum-scroll').directive('scrollable',
  function($timeout, $window, $document) {

    return {
      restrict : 'AE',
      scope : {
        scrollToPageTime : '@' || 400,
        currPageX : '=',
        currPageY : '=',
        currY : '=',
        currX : '=' ,
        isMaxX : '=',
        isMinX : '=',
        isMaxY : '=',
        isMinY : '=',
        onRefresh: '&',
        onBeforeScrollStart: '&',
        onScrollStart: '&',
        onBeforeScrollMove: '&',
        onScrollMove: '&',
        parameters : '@',
        onBeforeScrollEnd: '&',
        onScrollEnd: '&',
        onTouchEnd: '&',
        onDestroy: '&',
        onZoomStart: '&',
        onZoom: '&',
        onZoomEnd: '&'
      },
      transclude : true,
      template : '<div class="scroller" ng-transclude></div>',
      link : function(scope, element, attrs) {
      // add stying
        var style =
        '.inline-flex {' +
        '  display: -webkit-inline-flex;' +
        '  display: -moz-inline-flex;' +
        '  display: -ms-inline-flexbox;' +
        '  display: -ms-inline-flex;' +
        '  display: inline-flex;' +
        '  display: -webkit-inline-box;' +
        '  overflow: hidden;' +
        '  white-space: nowrap;' +
        '}' +
        '.inline-flex > * {' +
        '  display: block;' +
        '}';
        var head = angular.element($document[0].head);
        head.append('<style type="text/css">' + style + '</style>');
        // apply some necessary styling 
        element.css('overflow', 'hidden');
        element.css('position', 'relative');
        var scroller = angular.element(element.children('.scroller')[0]);

      //define handlers
        var handleScrollEnd = function() {
          $timeout(function(){
          // custom scrollend callback to intercept 'scroll' callback
            if (angular.isDefined(scope.currY)) {
              scope.currY = iScrollInstance.y;
            }
            if (angular.isDefined(scope.isMaxY)) {
              scope.isMaxY = (iScrollInstance.y <= iScrollInstance.maxScrollY);
            }
            if (angular.isDefined(scope.isMinY)) {
              scope.isMinY = (iScrollInstance.y >= 0);
            }
            if (angular.isDefined(scope.currX)) {
              scope.currX = iScrollInstance.x;
            }
            if (angular.isDefined(scope.isMaxX)) {
              scope.isMaxX = (iScrollInstance.x <= iScrollInstance.maxScrollX);
            }
            if (angular.isDefined(scope.isMinX)) {
              scope.isMinX = (iScrollInstance.x >= 0);
            }

            if (angular.isDefined(iScrollInstance.currentPage)) {
              if (angular.isDefined(scope.currPageY)) {
                scope.currPageY = iScrollInstance.currentPage.pageY;
              }
              if (angular.isDefined(scope.currPageX)) {
                scope.currPageX = iScrollInstance.currentPage.pageX;
              }
            }
            var state = {
              pageX: angular.isDefined(iScrollInstance.currentPage) ?
              iScrollInstance.currentPage.pageX : undefined,
              pageY: angular.isDefined(iScrollInstance.currentPage) ?
              iScrollInstance.currentPage.pageY : undefined,
              X: iScrollInstance.x,
              Y: iScrollInstance.y
            };
            scope.onScrollEnd(state);
          });
        };

        var iScrollInstance;
        var scrollToPageX = function (pageX) {
          if (angular.isDefined(iScrollInstance.pages) &&
            iScrollInstance.pages.length  !== 0 && angular.isDefined(pageX)) {
            iScrollInstance.goToPage(pageX, 0, scope.scrollToPageTime);
          }
        };
        var scrollToPageY = function (pageY) {
          if (angular.isDefined(iScrollInstance.pages) &&
            iScrollInstance.pages.length !== 0 && angular.isDefined(pageY)) {
            iScrollInstance.goToPage(0, pageY, scope.scrollToPageTime);
          }
        };
        var scrollToY = function (Y) {
          if (angular.isDefined(Y)) {
            iScrollInstance.scrollTo(0, Y, scope.scrollToPageTime);
          }
        };
        var scrollToX = function (newVal) {
          if (angular.isDefined(newVal)) {
            iScrollInstance.scrollTo(newVal, 0, scope.scrollToPageTime);
          }
        };

        var supportsOrientationChange = 'onorientationchange' in $window,
            orientationEvent = supportsOrientationChange ?
                'orientationchange' : 'resize';
        var scr = {};
        /* register for changes */
        var orientationEventListener = function orientationEventListener() {
          if (scr.width !== screen.width || scr.height !== screen.height) {
            scr = {'width' : screen.width, 'height' : screen.height};
            if (angular.isDefined(iScrollInstance)) {
              iScrollInstance.refresh();
            }
          }
        };

        // destroy iScrollInstance and cleaning up event handlers
        var _destroy = function() {
          if (angular.isDefined(iScrollInstance)) {
            $window.removeEventListener(orientationEvent, orientationEventListener, false);

            iScrollInstance.destroy();
            iScrollInstance = undefined;
          }
        };

        // initializer
        var _init = function() {
          _destroy();
          iScrollInstance = new IScroll(element[0], scope.iscrollParameters);
          // attach 'on'-callbacks
          for (var onMethod in scope) {
            if ((onMethod.indexOf('on') !== -1) &&
              scope.hasOwnProperty(onMethod) &&
              angular.isFunction(scope[onMethod])) {
              var event = onMethod.substring(2).charAt(0).toLowerCase() +
            onMethod.substring(2).slice(1);
              if (event === 'scrollEnd') { continue; }
              iScrollInstance.on(event, scope[onMethod]);
            }
          }

          iScrollInstance.on('scrollEnd', handleScrollEnd);
          scope.$watch('currPageY', scrollToPageY);
          scope.$watch('currPageX', scrollToPageX);
          scope.$watch('currY', scrollToY);
          scope.$watch('currX', scrollToX);

          // refresh the scroller on orientation change for mobile
          $window.addEventListener(orientationEvent, orientationEventListener, false);

          /* make sure to free memory if scrollable element is
          * destroyed (avoid memleaking)*/
          element.bind('$destroy', function() {
            if (angular.isDefined(iScrollInstance)) {
              _destroy();
            }
          });
        };

        attrs.$observe('parameters', function(val) {
          // parse the JSON string
          if (angular.isString(val)) {
            scope.iscrollParameters = angular.fromJson(val);
          }
          else {
            scope.iscrollParameters = val;
          }
          scope.scrollX = ('scrollX' in scope.iscrollParameters &&
            scope.iscrollParameters.scrollX);
          // fix for automatic horizontal scroll
          if (scope.scrollX){
            scroller.addClass('inline-flex');
          }
          // do not prevent Clicks into nested elements by default
          scope.iscrollParameters.click = angular.isDefined(val.click) ? val.click : true;
        });

        var _refresh = function(nVal) {
            if (angular.isDefined(nVal)) {
              $timeout(function(){
                if (angular.isDefined(iScrollInstance)) {
                  iScrollInstance.destroy();
                }
                _init();
                if (angular.isDefined(iScrollInstance.pages) &&
                  iScrollInstance.pages.length > 0) {
                  scrollToPageX(nVal.currPageX);
                  scrollToPageY(nVal.currPageY);
                }
              });
            }
          };

        // refresh on content change
        scope.$watch(function () { return scroller[0].offsetHeight; },
          _refresh
          );
        scope.$watch(function () { return scroller[0].offsetWidth; },
          _refresh
          );

        // initial refresh
        _refresh();
      }
    };
  });

})(angular);


demo/app/scripts/controllers/main.js
'use strict';

angular.module('demoApp')
  .controller('MainCtrl', function ($scope, $log) {
    $scope.currX = 0;
    $scope.currY = 0;
    $scope.currPageX = 0;
    $scope.currPageY = 0;
    $scope.isMaxY = false;
    $scope.isMinY = false;
    $scope.awesomeThings = [
      '01 scroll me down',
      '02 HTML5 Boilerplate',
      '03 AngularJS',
      '04 Karma',
      '05 iScroll',
      '06 momentum',
      '07 Demo',
      '08 Github',
      '09 HTML5',
      '10 Javascript',
      '11 -- and again --',
      '12 HTML5 Boilerplate',
      '13 AngularJS',
      '14 Karma',
      '15 iScroll',
      '16 momentum',
      '17 Demo',
      '18 Github',
      '19 HTML5',
      '20 Javascript',
      '21 -- and again --',
      '22 HTML5 Boilerplate',
      '23 AngularJS',
      '24 Karma',
      '25 iScroll',
      '26 momentum',
      '27 Demo',
      '28 Github',
      '29 HTML5',
      '30 Javascript',
      '31 -- and again --',
      '32 HTML5 Boilerplate',
      '33 AngularJS',
      '34 Karma',
      '35 iScroll',
      '36 momentum',
      '37 Demo',
      '38 Github',
      '39 HTML5',
      '40 Javascript'
    ];
    // setup the scroller
    $scope.iScrollParameters = {vScroll:true, scrollbars: true, snap : 'li', click: false};
    // setup the logger
    $scope.logger = [];
    $scope.log = function (msg) {
      var _msg = Date.now() + ' : ' + msg;
      $scope.logger.splice(0, 0, _msg);
      $log.log(_msg);
    };

    $scope.clicked= function (item) {
      $scope.log('clicked: ' + item);
      $log.log('clicked: ' + item);
    };
    $scope.logcallback = function(pageX, pageY, X, Y) {
      $scope.log('From callback: pageX ' + pageX + ' pageY ' + pageY +
          ' X ' + X, ' Y ' + Y);
    };
    $scope.$watch('currPageY', function() {
      $scope.log('From scope: pageX ' + $scope.currPageX + ' pageY ' +
          $scope.currPageY);
    });
    $scope.$watch('currY', function() {
      $scope.log('From scope: X ' + $scope.currX + ' Y ' +
          $scope.currY );
    });

    $scope.$watch('isMinY', function() {
      if ($scope.isMinY) {
        $scope.log('MinY is reached');
      }
    });
    $scope.$watch('isMaxY', function() {
      if ($scope.isMaxY) {
        $scope.log('MaxY is reached');
      }
    });
    // scope methods
    $scope.scrollToPage = function(page) {
      $scope.currPageY = page;
    };
    $scope.append = function(item) {
      $scope.log('pushing ' + item + ' to back of the list of awesome things');
      $scope.awesomeThings.push(item);
    };
    $scope.prepend = function(item) {
      $scope.log('pushing ' + item + ' to front of the list of awesome things');
      $scope.awesomeThings.splice(0, 0, item);
    };
  });
